File: src/main/java/burst/pool/miners/MinerMaths.java
package burst.pool.miners;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigInteger;

public class MinerMaths {
    private final Logger logger = LoggerFactory.getLogger(MinerMaths.class);

    private static final long GenesisBaseTarget = 18325193796L;
    private final double[] alphas;

    public MinerMaths(int nAvg, int nMin) {
        alphas = new double[nAvg];
        for (int i = 0; i < nAvg; i++) {
            if (i < nMin-1) {
                alphas[i] = 0d;
            } else {
                double nConf = i + 1;
                alphas[i] = 1d - ((double)nAvg-nConf)/ nConf *Math.log(nAvg/((double)nAvg-nConf));
            }
        }
        alphas[nAvg-1] = 1d;
    }

    public double estimatedEffectivePlotSize(int originalNConf, int nConf, BigInteger hitSum) {
        if (hitSum.compareTo(BigInteger.ZERO) == 0) {
            throw new ArithmeticException();
        }
        double plotSize =  alpha(originalNConf) * 240d * (((double)nConf)-1d) / (hitSum.divide(BigInteger.valueOf(GenesisBaseTarget)).longValue());
        if (Double.isInfinite(plotSize) || Double.isNaN(plotSize)) {
            logger.debug("Calculated impossible plot size. originalNConf: " + originalNConf + ", nConf: " + nConf + ", hitSum: " + hitSum);
            throw new ArithmeticException();
        }
        return plotSize;
    }

    private double alpha(int nConf) {
        if (nConf == 0) {
            return 0d;
        }
        if (alphas.length < nConf) {
            return 1d;
        }
        return alphas[nConf-1];
    }
}

File: src/main/kotlin/burst/pool/miners/MinerMaths.kt
package burst.pool.miners

import org.slf4j.LoggerFactory
import java.math.BigInteger

class MinerMaths(nAvg: Int, nMin: Int) {
    private val logger = LoggerFactory.getLogger(MinerMaths::class.java)
    private val alphas: DoubleArray
    fun estimatedEffectivePlotSize(originalNConf: Int, nConf: Int, hitSum: BigInteger): Double {
        if (hitSum.compareTo(BigInteger.ZERO) == 0) {
            throw ArithmeticException()
        }
        val plotSize: Double = alpha(originalNConf) * 240.0 * (nConf.toDouble() - 1.0) / hitSum.divide(BigInteger.valueOf(GenesisBaseTarget)).toLong()
        if (java.lang.Double.isInfinite(plotSize) || java.lang.Double.isNaN(plotSize)) {
            logger.debug("Calculated impossible plot size. originalNConf: $originalNConf, nConf: $nConf, hitSum: $hitSum")
            throw ArithmeticException()
        }
        return plotSize
    }

    private fun alpha(nConf: Int): Double {
        if (nConf == 0) {
            return 0.0
        }
        return if (alphas.size < nConf) {
            1.0
        } else alphas[nConf - 1]
    }

    companion object {
        private const val GenesisBaseTarget = 18325193796L
    }

    init {
        alphas = DoubleArray(nAvg)
        for (i in 0 until nAvg) {
            if (i < nMin - 1) {
                alphas[i] = 0.0
            } else {
                val nConf = i + 1.toDouble()
                alphas[i] = 1.0 - (nAvg.toDouble() - nConf) / nConf * Math.log(nAvg / (nAvg.toDouble() - nConf))
            }
        }
        alphas[nAvg - 1] = 1.0
    }
}
